### ISSUE TITLE
History state after mutation not being updated in tests

### ISSUE THREAD

**Issue by bsides** [tannerlinsley/react-query#3184]
### Describe the bug

TL;DR: After a mutation succeeds, if I send a `history.push(url, STATE)` the STATE never gets updated in the same page I am.

Quick example:

```javascript
// current url: /someurl/1

function SomeComponent() {
  const history = useHistory()
  const location = useLocation()
  const mutation = useMutation(somefetch, {
    onSuccess: (data) => history.push(`/someurl/${data.id}`, { THIS_DATA_HERE: 'SOMETHING' })
  })

  return (
    <>
      {location?.state ? <p>id: {location.state.THIS_DATA_HERE}</p> : null}
      <button type="button" onClick={() => mutation.mutate({something: 'somedata'})}>create</button>
    </>
  )
}
```

After running the mutation, the block that sends the user to `/someurl/${data.id}` isnt getting the state from `{ THIS_DATA_HERE: 'SOMETHING' }` in tests. In the browser it works out fine.

I did some reproductions in codesandbox to be sure it was an error with `react-query` itself and not jest, jest-dom, testing-library, react or react-dom. Although I couldn't precise it 100%, I did 2 codes: one with react-query and other simulating the promise after clicking the button but without react-query, and the 2nd one passes.

With react-query:
https://codesandbox.io/s/thirsty-wind-duv4m?file=/src/App.tsx

Without react-query:
https://codesandbox.io/s/qdoie?file=/src/App.tsx

### Your minimal, reproducible example

https://codesandbox.io/s/thirsty-wind-duv4m?file=/src/App.tsx

### Steps to reproduce

1. Go to the link
2. Click "Go to User 1"
3. Click "Add User 2"

### Expected behavior

In the browser you will notice everything works fine, but in the test it never passes because it never gets the state from the history.

It does work without react-query, though.

### How often does this bug happen?

Every time

### Platform

- Browser: Chrome
- OS: MacOS
- ran reproduction in chrome as well

### react-query version

3.34.7

### TypeScript version

4.5.4

**Comment by TkDodo** [comment]
@bsides I've re-added react-query mutations to your sandbox and it seems to work fine: https://codesandbox.io/s/friendly-brahmagupta-1dz7v?file=/src/App.test.js

maybe you can add more code until it breaks again? maybe you removed something else, too? Maybe it's because of the actual network request or so, because I've just used your promise stub for now ...

--- 
from: https://github.com/TkDodo/blog-comments/issues/22#issuecomment-1013098275

sorry, I didn't see that you've opened an issue. let's try to keep the discussion here :)

**Comment by TkDodo** [comment]
closing because my sandbox shows that it works even with react-query, so it must be something else. feel free to reopen if you think there is something we can do in react-query for this issue.
